internal/receiver/smartagentreceiver/receiver.go

```diff
@@ -135,5 +149,57 @@ func (r *Receiver) createMonitor(monitorType string) (interface{}, error) {
 		output.AddExtraDimension(k, v)
 	}
 
-	return monitor, nil
+	// Configure SmartAgentConfigProvider to gather any global config overrides and
+	// set required envs.
+	configureEnvironmentOnce.Do(func() {
+		setUpSmartAgentConfigProvider(extensions)
+		setUpEnvironment()
+	})
+
+	// Note, that this receiver has to configure main collectd even for collectd/custom monitor,
+	// despite the fact that, that monitor stands up its own instance of collectd to prevent this
+	// panic "Main collectd instance should not be accessed before being configured".
+	if r.config.monitorConfig.MonitorConfigCore().IsCollectdBased() {
+		configureCollectdOnce.Do(func() {
+			r.logger.Info("Configuring collectd")
+			err = configureMainCollectd(saConfigProvider.CollectdConfig())
+		})
+	}
+
+	return monitor, err
+}
+
+func configureMainCollectd(collectdConfig config.CollectdConfig) error {
+	return collectd.ConfigureMainCollectd(&collectdConfig)
+}
+
+func setUpSmartAgentConfigProvider(extensions map[configmodels.Extension]component.ServiceExtension) {
+	// If smartagent extension is not configured, use the default config.
+	f := smartagentextension.NewFactory()
+	defaultCfg := f.CreateDefaultConfig().(smartagentextension.SmartAgentConfigProvider)
+	saConfigProvider = defaultCfg
+
+	// Do a lookup for any smartagent extensions to pick up common collectd options
+	// to be applied across instances of the receiver.
+	for c := range extensions {
+		if c.Type() != f.Type() {
+			continue
+		}
+
+		cfgProvider, ok := c.(smartagentextension.SmartAgentConfigProvider)
+		if !ok {
+			continue
+		}
+		saConfigProvider = cfgProvider
+
+		// If there are multiple extensions configured, pick the first one. Ideally,
+		// there would only be one extension.
+		break
```

Comment:
Would it be better experience to iterate through all of them and log a warning when multiples are provided?

Reply:
Updated.